fedimint-server/src/net/peers.rs

```diff
@@ -214,189 +138,139 @@ where
-            let new_connection = tokio::select! {
-                maybe_msg = listener.next() => { maybe_msg },
-                () = &mut shutdown_rx => { break; },
-            };
+            .expect("Could not bind to port");
+
+        task_group.spawn("handle-incoming-p2p-connections", move |handle| async move {
+            loop {
+                tokio::select! {
+                    connection = listener.next() => {
+                        match connection.expect("Listener closed") {
+                            Ok((peer, connection)) => {
+                                if connection_senders
+                                    .get_mut(&peer)
+                                    .expect("Authenticating connectors dont return unknown peers")
+                                    .send(connection)
+                                    .await
+                                    .is_err()
+                                {
+                                    break;
+                                }
+                            },
+                            Err(e) => {
+                                warn!(target: LOG_NET_PEER, "Error while opening incoming connection: {}", e);
+                            }
+                        }
+                    },
+                    () = &mut handle.make_shutdown_rx() => { break },
+                };
+            }
 
-            let (peer, connection) = match new_connection.expect("Listener closed") {
-                Ok(connection) => connection,
-                Err(e) => {
-                    warn!(target: LOG_NET_PEER, mint = ?cfg.identity, err = %e, "Error while opening incoming connection");
-                    continue;
-                }
-            };
+            info!(target: LOG_NET_PEER, "Shutting down task listening for p2p connections");
+        });
 
-            let err = connection_senders
-                .get_mut(&peer)
-                .expect("Authenticating connectors should not return unknown peers")
-                .send(connection)
-                .await
-                .is_err();
-
-            if err {
-                warn!(
-                    target: LOG_NET_PEER,
-                    ?peer,
-                    "Could not send incoming connection to peer io task (possibly banned)"
-                );
-            }
-        }
+        ReconnectPeerConnections { connections }
     }
-    pub fn send_sync(&self, msg: &T, recipient: Recipient) {
+}
+
+#[async_trait]
+impl<M> IPeerConnections<M> for ReconnectPeerConnections<M>
+where
+    M: std::fmt::Debug + Serialize + DeserializeOwned + Clone + Unpin + Send + Sync + 'static,
+{
+    async fn send(&mut self, recipient: Recipient, msg: M) {
         match recipient {
             Recipient::Everyone => {
                 for connection in self.connections.values() {
-                    connection.send(msg.clone());
+                    connection.send(msg.clone()).await;
                 }
```

Comment:
Now I'm wondering if implementing "send to all" with normal `send` instead of `send_all` calls was ever safe.